### PR TITLE
Use more optimal Ord implementation for integers

### LINKED ISSUE #63758: Implementation of Ord for integers is suboptimal
The current implementation of `Ord::cmp` for integral types results in less than optimal code. Currently the implementation looks something like this:

```rust                
fn cmp(&self, other: &$t) -> Ordering {
    if *self == *other { Equal }
    else if *self < *other { Less }
    else { Greater }
}
```

This results in the following IR:

```llvm
  %2 = icmp eq i64 %0, %1
  %3 = icmp ult i64 %0, %1
  %..i = select i1 %3, i8 -1, i8 1
  %_0.0.i = select i1 %2, i8 0, i8 %..i
  ret i8 %_0.0.i
```

which, on x86, becomes

```asm
	cmpq	%rsi, %rdi
	setae	%al
	cmpq	%rsi, %rdi
	je	.LBB0_1
	addb	%al, %al
	addb	$-1, %al
        ; exit (retq)
    .LBB0_1:
	xorl	%eax, %eax
        ; exit (retq)
```

where the critical path looks like this (courtesy of `llvm-mca -mcpu=broadwell`):

```
              Instruction                                 Dependency Information
 +----< 0.    cmpq	%rsi, %rdi
 +----> 1.    setae	%al                               ## REGISTER dependency:  %flags
 |      2.    cmpq	%rsi, %rdi
 |      3.    je	.LBB0_1
 +----> 4.    addb	%al, %al                          ## REGISTER dependency:  %al
 +----> 5.    addb	$-1, %al                          ## REGISTER dependency:  %al
 |      6.    ; exit
 |      7.    xorl	%eax, %eax
 |      8.    ; exit
 |
 |    < loop carried > 
 |
 +----> 2.    cmpq	%rsi, %rdi                        ## RESOURCE interference:  BWPort1 [ probability: 25% ]
```

---

if the implementation instead was:

```rust
    if a < b {
        std::cmp::Ordering::Less
    } else if a > b {
        std::cmp::Ordering::Greater
    } else {
        std::cmp::Ordering::Equal
    }
```

the IR would become

```llvm
  %0 = icmp ult i64 %a, %b
  %1 = icmp ugt i64 %a, %b
  %. = zext i1 %1 to i8
  %_0.0 = select i1 %0, i8 -1, i8 %.
  ret i8 %_0.0
```

which in turn would compile down to

```asm
	movb	$-1, %al
	cmpq	%rsi, %rdi
	seta	%cl
	jb	.LBB1_2
	movl	%ecx, %eax
    .LBB1_2:
        ; exit (retq)
```

for which the critical path (as expected) only becomes visible on a 2nd iteration:

```
              Instruction                                 Dependency Information
 +----< 2.    seta	%cl
 |
 |    < loop carried > 
 |
 |      0.    movb	$-1, %al
 +----> 1.    cmpq	%rsi, %rdi                        ## RESOURCE interference:  BWPort6 [ probability: 97% ]
 +----> 2.    seta	%cl                               ## REGISTER dependency:  %flags
 |      3.    jb	.LBB1_2
 +----> 4.    movl	%ecx, %eax                        ## REGISTER dependency:  %cl
```

llvm-mca reports that the reciprocal throughput for the improved version is 1.5 or lower (getting as low as 1.3 on znver1; lower is better for reciprocal throughput) for various x86 architectures whereas the old code always exceeds 2.0, reaching 3.0 on `core2`.

### PR DESCRIPTION
**Pull request by tesuji**
Closes #63758
r? @nagisa 

### Compare results

([godbolt link](https://godbolt.org/z/dsbczy))

Old assembly:
```asm
example::cmp1:
  mov eax, dword ptr [rdi]
  mov ecx, dword ptr [rsi]
  cmp eax, ecx
  setae dl
  add dl, dl
  add dl, -1
  xor esi, esi
  cmp eax, ecx
  movzx eax, dl
  cmove eax, esi
  ret
```

New assembly:
```asm
example::cmp2:
  mov eax, dword ptr [rdi]
  xor ecx, ecx
  cmp eax, dword ptr [rsi]
  seta cl
  mov eax, 255
  cmovae eax, ecx
  ret
```

Old llvm-mca statistics:
```
Iterations:        100
Instructions:      1100
Total Cycles:      243
Total uOps:        1300

Dispatch Width:    6
uOps Per Cycle:    5.35
IPC:               4.53
Block RThroughput: 2.2
```

New llvm-mca statistics:
```
Iterations:        100
Instructions:      700
Total Cycles:      217
Total uOps:        1100

Dispatch Width:    6
uOps Per Cycle:    5.07
IPC:               3.23
Block RThroughput: 1.8
```

### CODE REVIEW

**Comment by matthiaskrgr** [comment]
Could you add a comment explaining that the ordering is performance critical here? (perhaps with a link to the original ticket)
This should make sure that it es not changed back to something slower by accident.

**Comment by hellow554** [comment]
What about adding a assembly testcase as well to prevent regressions (e.g. due to other optimizations?)

**Comment by lzutao** [comment]
> What about adding a assembly testcase as well to prevent regressions (e.g. due to other optimizations?)

I don't know how. Could you give a mentor?

**Comment by hellow554** [comment]
@lzutao You can take a look at https://rust-lang.github.io/rustc-guide/tests/intro.html and the codegen test cases in https://github.com/rust-lang/rust/tree/master/src/test/codegen especially at 
https://github.com/rust-lang/rust/blob/bea0372a1a7a31b81f28cc4d9a83a2dc9a79d008/src/test/codegen/float_math.rs#L1-L50 I guess

**Comment by lzutao** [comment]
Hi @hellow554, I wrote a simple test for this. But honestly, I don't know how to match
on this llvm IR output:

```asm
; ModuleID = 'integer_cmp.3a1fbbbh-cgu.0'
source_filename = "integer_cmp.3a1fbbbh-cgu.0"
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

; core::cmp::impls::<impl core::cmp::Ord for i64>::cmp
; Function Attrs: inlinehint nonlazybind uwtable
define internal i8 @"_ZN4core3cmp5impls48_$LT$impl$u20$core..cmp..Ord$u20$for$u20$i64$GT$3cmp17h54b2deb40875460aE"(i64* noalias readonly align 8 dereferenceable(8) %self, i64* noalias readonly align 8 dereferenceable(8) %other) unnamed_addr #0 {
start:
  %_0 = alloca i8, align 1
  %0 = load i64, i64* %self, align 8
  %1 = load i64, i64* %other, align 8
  %2 = icmp slt i64 %0, %1
  br i1 %2, label %bb2, label %bb1

bb1:                                              ; preds = %start
  %3 = load i64, i64* %self, align 8
  %4 = load i64, i64* %other, align 8
  %5 = icmp sgt i64 %3, %4
  br i1 %5, label %bb4, label %bb3

bb2:                                              ; preds = %start
  store i8 -1, i8* %_0, align 1
  br label %bb6

bb3:                                              ; preds = %bb1
  store i8 0, i8* %_0, align 1
  br label %bb5

bb4:                                              ; preds = %bb1
  store i8 1, i8* %_0, align 1
  br label %bb5

bb5:                                              ; preds = %bb3, %bb4
  br label %bb6

bb6:                                              ; preds = %bb5, %bb2
  %6 = load i8, i8* %_0, align 1, !range !1
  ret i8 %6
}

; core::cmp::impls::<impl core::cmp::Ord for u32>::cmp
; Function Attrs: inlinehint nonlazybind uwtable
define internal i8 @"_ZN4core3cmp5impls48_$LT$impl$u20$core..cmp..Ord$u20$for$u20$u32$GT$3cmp17h1dd39efa68a3677aE"(i32* noalias readonly align 4 dereferenceable(4) %self, i32* noalias readonly align 4 dereferenceable(4) %other) unnamed_addr #0 {
start:
  %_0 = alloca i8, align 1
  %0 = load i32, i32* %self, align 4
  %1 = load i32, i32* %other, align 4
  %2 = icmp ult i32 %0, %1
  br i1 %2, label %bb2, label %bb1

bb1:                                              ; preds = %start
  %3 = load i32, i32* %self, align 4
  %4 = load i32, i32* %other, align 4
  %5 = icmp ugt i32 %3, %4
  br i1 %5, label %bb4, label %bb3

bb2:                                              ; preds = %start
  store i8 -1, i8* %_0, align 1
  br label %bb6

bb3:                                              ; preds = %bb1
  store i8 0, i8* %_0, align 1
  br label %bb5

bb4:                                              ; preds = %bb1
  store i8 1, i8* %_0, align 1
  br label %bb5

bb5:                                              ; preds = %bb3, %bb4
  br label %bb6

bb6:                                              ; preds = %bb5, %bb2
  %6 = load i8, i8* %_0, align 1, !range !1
  ret i8 %6
}

; Function Attrs: nonlazybind uwtable
define i8 @cmp_signed(i64, i64) unnamed_addr #1 {
start:
  %b = alloca i64, align 8
  %a = alloca i64, align 8
  store i64 %0, i64* %a, align 8
  store i64 %1, i64* %b, align 8
; call core::cmp::impls::<impl core::cmp::Ord for i64>::cmp
  %2 = call i8 @"_ZN4core3cmp5impls48_$LT$impl$u20$core..cmp..Ord$u20$for$u20$i64$GT$3cmp17h54b2deb40875460aE"(i64* noalias readonly align 8 dereferenceable(8) %a, i64* noalias readonly align 8 dereferenceable(8) %b), !range !1
  br label %bb1

bb1:                                              ; preds = %start
  ret i8 %2
}

; Function Attrs: nonlazybind uwtable
define i8 @cmp_unsigned(i32, i32) unnamed_addr #1 {
start:
  %b = alloca i32, align 4
  %a = alloca i32, align 4
  store i32 %0, i32* %a, align 4
  store i32 %1, i32* %b, align 4
; call core::cmp::impls::<impl core::cmp::Ord for u32>::cmp
  %2 = call i8 @"_ZN4core3cmp5impls48_$LT$impl$u20$core..cmp..Ord$u20$for$u20$u32$GT$3cmp17h1dd39efa68a3677aE"(i32* noalias readonly align 4 dereferenceable(4) %a, i32* noalias readonly align 4 dereferenceable(4) %b), !range !1
  br label %bb1

bb1:                                              ; preds = %start
  ret i8 %2
}

attributes #0 = { inlinehint nonlazybind uwtable "probe-stack"="__rust_probestack" "target-cpu"="x86-64" }
attributes #1 = { nonlazybind uwtable "probe-stack"="__rust_probestack" "target-cpu"="x86-64" }

!llvm.module.flags = !{!0}

!0 = !{i32 2, !"RtLibUseGOT", i32 1}
!1 = !{i8 -1, i8 2}
```

File: src/test/codegen/integer-cmp.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+// This is test for more optimal Ord implementation for integers.
+// See <https://github.com/rust-lang/rust/issues/63758> for more info.
+
+// compile-flags: -C no-prepopulate-passes -C opt-level=3
```
Comment (nagisa) [inline src/test/codegen/integer-cmp.rs]: You’ll want to ensure this test is compiled with optimisations. Remove `-Cno-prepopulate-passes`.

Comment (lzutao) [reply src/test/codegen/integer-cmp.rs]: Nice. Thank you.

**Comment by lzutao** [comment]
The CI is green.

**Pull request merged**